src/Language/DifferentialDatalog/Parse.hs

```diff
@@ -250,14 +276,13 @@ parseLetStatement = withPos $ (LetStatement nopos) <$ reserved "let"
                                                   <*> commaSep parseAssignment
                                                   <*> (reserved "in" *> statement)
 
-parseInsertStatement = withPos $ (InsertStatement nopos) <$> relIdent
-                                                         <*> (parens $ commaSep expr)
+parseInsertStatement = withPos $ (InsertStatement nopos) <$> atom
 
-rule = withPos $ Rule nopos <$>
-                 (commaSep1 atom) <*>
-                 (option [] (reservedOp ":-" *> commaSep rulerhs)) <* dot
+rule = Rule nopos <$>
+       (commaSep1 atom) <*>
+       (option [] (reservedOp ":-" *> commaSep rulerhs)) <* dot
 
-rulerhs =  (do _ <- try $ lookAhead $ (optional $ reserved "not") *> relIdent *> symbol "("
+rulerhs =  (do _ <- try $ lookAhead $ (optional $ reserved "not") *> relIdent *> (symbol "(" <|> symbol "[")
```

Comment:
I was expecting a closed "]" somewhere.

Reply:
This is a lookahead, it does not consume any input; just checks that we are looking at an atom before actually parsing it, so seeing the opening bracket is enough